### PR TITLE
Add remote configuration repository

### PR DESCRIPTION
**Pull request by lloeki**
**What does this PR do?**

Add configuration repository, using PORO from #2687

**Motivation**

**Additional Notes**

This is an initial basic implementation. It is notably missing:

- storing digests for each bit of content
- reporting a summary of which paths have been updated and their corresponding data for taking action on the newly received configuration
- error tracking and reporting
- a content caching mechanism and cache state reporting

These will be added progressively.

**How to test the change?**

```
  require 'datadog/core/remote/configuration/path'
  require 'datadog/core/remote/configuration/target'
  require 'datadog/core/remote/configuration/content'

  path = Datadog::Core::Remote::Configuration::Path.parse('datadog/42/PRODUCT/foo/config')
  content = Datadog::Core::Remote::Configuration::Content.parse({path: 'datadog/42/PRODUCT/foo/config', content: StringIO.new('Lorem ipsum')})
  target = Datadog::Core::Remote::Configuration::Target.parse({'length' => 11, 'hashes' => {'sha256' => 'a9a66978f378456c818fb8a3e7c6ad3d2c83e62724ccbdea7b36253fb8df5edd'}})

  path2 = Datadog::Core::Remote::Configuration::Path.parse('datadog/42/PRODUCT/bar/config')
  content2 = Datadog::Core::Remote::Configuration::Content.parse({path: 'datadog/42/PRODUCT/bar/config', content: StringIO.new('Lorem ipsum')})
  target2 = Datadog::Core::Remote::Configuration::Target.parse({'length' => 11, 'hashes' => {'sha256' => 'a9a66978f378456c818fb8a3e7c6ad3d2c83e62724ccbdea7b36253fb8df5edd'}})

  content3 = Datadog::Core::Remote::Configuration::Content.parse({path: 'datadog/42/PRODUCT/foo/config', content: StringIO.new('dolor sit amet')})
  target3 = Datadog::Core::Remote::Configuration::Target.parse({'length' => 11, 'hashes' => {'sha256' => 'aa8311d08b68a5fdda55ad0947fff3c5a4b2397f5f766e9c9a79f4a5486c633c'}})

  require 'datadog/core/remote/configuration/repository'

  txn = Datadog::Core::Remote::Configuration::Repository::Transaction.new
  txn.delete(path)
  txn.operations

  txn = Datadog::Core::Remote::Configuration::Repository::Transaction.new
  txn.insert(path, target, content)
  txn.operations

  txn = Datadog::Core::Remote::Configuration::Repository::Transaction.new
  txn.update(path, target, content)
  txn.operations

  txn = Datadog::Core::Remote::Configuration::Repository::Transaction.new
  txn.set(opaque_backend_state: 'OPAQUE_DATA')
  txn.operations

  txn = Datadog::Core::Remote::Configuration::Repository::Transaction.new
  txn.set(targets_version: 42)
  txn.operations

  repository = Datadog::Core::Remote::Configuration::Repository.new
  repository.contents
  repository.root_version
  repository.targets_version

  repository.transaction do |repo, txn|
    txn.insert(path, target, content)
    txn.set(opaque_backend_state: 'OPAQUE_DATA', targets_version: target_map.version)
  end

  repository.contents
  repository.root_version
  repository.targets_version
  repository[path]

  repository.transaction do |repo, txn|
    txn.insert(path2, target2, content2)
  end

  repository[path2]
  repository.contents

  repository.transaction do |repo, txn|
    txn.insert(path, target, content)
  end

  repository.contents

  repository.transaction do |repo, txn|
    txn.update(path, target3, content3)
  end

  repository[path]

  repository.transaction do |repo, txn|
    txn.delete(path)
  end

  repository.contents
  repository[path]

  repository.transaction do |repo, txn|
    txn.update(path, target, content)
  end

  repository[path]
```

### CODE REVIEW

**Comment by lloeki** [comment]
Pending:

- specs
- RBS typing

**Comment by lloeki** [comment]
Rebased using `git rebase --onto origin/add-remote-config-poro 37eb3b095 add-remote-config-repository`

**Comment by lloeki** [comment]
Rebased using `git rebase --onto master 414a626^ add-remote-config-repository`

File: lib/datadog/core/remote/configuration/repository.rb [hunk 1]
```diff
@@ -91,10 +91,17 @@ def set(**options)
           end
 
           module Operation
-            class Delete
+            class Base
+              def apply(repository)
+                raise NoMethodError
+              end
+            end
```
Comment (lloeki) [inline lib/datadog/core/remote/configuration/repository.rb:103]: I presume this is for the sake of typing?

A better approach would be to [create a union type alias](https://github.com/ruby/rbs/blob/0dd11d38c11f55ead0a0d78bb9b19d2eaf3a2104/docs/syntax.md#type-alias-declaration) in `repository.rbs`:

```
module Operation
    ...
end

type operation = Operation::Insert | Operation::Update | Operation::Delete | Operation::Set
```

and then:

```
attr_reader operations: Array[operation]
```

**Review by lloeki — COMMENTED** [review]
Thanks for the specs and types! I have a suggestion concerning `Base`.

File: lib/datadog/core/remote/configuration/repository.rb (hunk 1 above)
Comment (lloeki) [reply lib/datadog/core/remote/configuration/repository.rb:103]: Or maybe [as an interface](https://github.com/ruby/rbs/blob/0dd11d38c11f55ead0a0d78bb9b19d2eaf3a2104/docs/syntax.md#interface-declaration), which would enforce `apply` being defined without repetition:

```
interface _Operation
  def apply: (Repository repository) -> void
end

module Operation
  class Insert
    include _Operation
  end
end
```

then:

```
attr_reader operations: Array[_Operation]
```

**Review by GustavoCaso — APPROVED** [review]

**Pull request merged**